Use `mut` less in dataflow analysis

nnethercote opened the pull request:
`&mut Analysis` is used a lot:
- In results visitors, even though only one visitor needs mutability.
- In all the `apply_*` methods, even though only one visitor needs mutability.

I've lost track of the number of times I've thought "why are these `mut` again?" and had to look through the code to remind myself. It's really unexpected, and most `Analysis` instances are immutable, because the `state` values are what get mutated.

This commit introduces `RefCell` in one analysis and one results visitor. This then lets another existing `RefCell` be removed, and a ton of `&mut Analysis` arguments become `&Analysis`. And then `Analysis` and `Results` can be recombined.

r? @cjgillot

nnethercote committed:
Make `ConstAnalysis::ecx` a `RefCell`.

Of the many dataflow analyses, `ConstAnalysis` is the only one that requires
the analysis be mutabile when used with `ResultsVisitor`. It's needed
because of the `ecx` field -- `ecx.intern_with_temp_alloc` is called
during visiting and it takes `&mut self`.

This commit changes `ConstAnalysis` to use interior mutability for the
`ecx` field. This is a bit annoying for `ConstAnalysis`, but it will
allow more immutability in `ResultsVisitor`, as seen in the next commit.

nnethercote committed:
Make `Analysis` immutable in `ResultsVisitor::visit_*` methods.

This makes sense -- you wouldn't expect that visiting the results of an
analysis would change the analysis itself.

nnethercote committed:
Use a `RefCell` in `MaybeRequiresStorage`.

This will let us make `Analysis` arguments in many other places
immutable, in the next commit.

nnethercote committed:
Make `Analysis` immutable in many more places.

The `state: A::Domain` value is the primary things that's modified when
performing an analysis. The `Analysis` impl is immutable in every case
but one (`MaybeRequiredStorage`) and it now uses interior mutability.

As well as changing many `&mut A` arguments to `&A`, this also:
- lets `CowMut` be replaced with the simpler `SimpleCow` in `cursor.rs`;
- removes the need for the `RefCell` in `Formatter`;
- removes the need for `MaybeBorrowedLocals` to impl `Clone`, because
  it's a unit type and it's now clear that its constructor can be used
  directly instead of being put into a local variable and cloned.

nnethercote committed:
Reorder args for `visit_results_in_block`.

Put `analysis` first, to match `apply_effects_in_range`.

hkBst reviewed (commented):
I'm not sure using more `RefCell` is to my taste...

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148165/23fced0fcc5e0ec260d25f04a8b78b269e5e90f0..e3b54c1f9d25a875aeea9da953af9f6d4ecb5c18)*

rust-timer commented:
Queued 7bd3a3001d02f884327e0ff6cd77d2ac00ac0c25 with parent 23fced0fcc5e0ec260d25f04a8b78b269e5e90f0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=23fced0fcc5e0ec260d25f04a8b78b269e5e90f0&end=7bd3a3001d02f884327e0ff6cd77d2ac00ac0c25).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~3.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

nnethercote committed:
Put `Analysis` back into `Results`.

`Results` used to contain an `Analysis`, but it was removed in #140234.
That change made sense because the analysis was mutable but the entry
states were immutable and it was good to separate them so the mutability
of the different pieces was clear.

Now that analyses are immutable there is no need for the separation,
lots of analysis+results pairs can be combined, and the names are going
back to what they were before:
- `Results` -> `EntryStates`
- `AnalysisAndResults` -> `Results`

cjgillot commented on compiler/rustc_mir_dataflow/src/framework/graphviz.rs:
```diff
@@ -206,12 +204,7 @@ where
-    // The `RefCell` is used because `<Formatter as Labeller>::node_label`
-    // takes `&self`, but it needs to modify the analysis. This is also the
-    // reason for the `Formatter`/`BlockFormatter` split; `BlockFormatter` has
-    // the operations that involve the mutation, i.e. within the `borrow_mut`.
```
Does this `Formatter`/`BlockFormatter` split still exist? Should it be removed?

nnethercote replied:
I think the comment was overstated. The RefCell was part of the reason for the separation, but not the only reason. `Formatter` and `BlockFormatter` both have numerous methods, and multiple `BlockFormatters` can exist for a single `Formatter`. I'm not even sure if they could be sensibly combined, and keeping them separate seems reasonable.

cjgillot commented:
Great!
Perf is neutral
@bors r+ rollup

bors commented:
:pushpin: Commit 87932397024a9053ca028adbba8315fc93cfa343 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 87932397024a9053ca028adbba8315fc93cfa343 -->
<!-- homu: {"type":"Approved","sha":"87932397024a9053ca028adbba8315fc93cfa343","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 87932397024a9053ca028adbba8315fc93cfa343 with merge b51dc8aa836c36c3561e86e0f2fec22db54f6186...
<!-- homu: {"type":"BuildStarted","head_sha":"87932397024a9053ca028adbba8315fc93cfa343","merge_sha":"b51dc8aa836c36c3561e86e0f2fec22db54f6186"} -->

Zalathar commented:
Yielding to enclosing rollup.

@bors retry
